File: src/main/java/jp/kentan/minecraft/neko_core/manager/AdvertisementManager.java
package jp.kentan.minecraft.neko_core.manager;

import jp.kentan.minecraft.neko_core.bridge.EconomyProvider;
import jp.kentan.minecraft.neko_core.component.AdvertiseFrequency;
import jp.kentan.minecraft.neko_core.component.Advertisement;
import jp.kentan.minecraft.neko_core.config.PlayerConfigProvider;
import jp.kentan.minecraft.neko_core.dao.AdvertisementDao;
import jp.kentan.minecraft.neko_core.event.AdvertisementEvent;
import jp.kentan.minecraft.neko_core.util.Util;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitScheduler;

import java.time.ZonedDateTime;
import java.util.*;

public class AdvertisementManager implements AdvertisementEvent {

    public final static String PREFIX = ChatColor.translateAlternateColorCodes('&', "&7[&d広告&7]&r ");

    private final BukkitScheduler SCHEDULER = Bukkit.getScheduler();
    private final List<Advertisement> AD_LIST = Collections.synchronizedList(new ArrayList<>());
    private final Map<Player, AdvertiseFrequency> PLAYER_MAP = Collections.synchronizedMap(new HashMap<>());
    private final Map<Player, ConfirmTask> CONFIRM_TASK_MAP = new HashMap<>();
    private final int PRICE_PER_DAY = 100;

    private final Plugin PLUGIN;
    private final AdvertisementDao DAO;
    private final EconomyProvider ECONOMY;
    private final PlayerConfigProvider PLAYER_CONFIG;

    private int mIndexAdListHigh   = 0;
    private int mIndexAdListMiddle = 0;
    private int mIndexAdListLow    = 0;

    public AdvertisementManager(Plugin plugin, EconomyProvider economyProvider, PlayerConfigProvider playerConfigProvider) {
        PLUGIN = plugin;
        ECONOMY = economyProvider;
        PLAYER_CONFIG = playerConfigProvider;

        DAO = new AdvertisementDao();

        syncLatestAds();

        Bukkit.getOnlinePlayers().forEach(p -> PLAYER_MAP.put(p, PLAYER_CONFIG.getAdvertiseFrequency(p.getUniqueId())));

        startBroadcastAsyncTask();
    }

    private void syncLatestAds() {
        SCHEDULER.runTaskAsynchronously(PLUGIN, () -> {
            AD_LIST.clear();
            AD_LIST.addAll(DAO.getAdvertisementList());
        });
    }

    private void startBroadcastAsyncTask() {
        final long INTERVAL_TICK = 20L * 60 * 5; //5m

        SCHEDULER.runTaskTimerAsynchronously(PLUGIN, () -> {
            AD_LIST.removeIf(ad -> ad.EXPIRE_DATE.isBefore(ZonedDateTime.now()));
            if (AD_LIST.isEmpty() || PLAYER_MAP.isEmpty()) {
                return;
            }

            if (AD_LIST.size() <= ++mIndexAdListHigh) {
                mIndexAdListHigh = 0;
            }
            Advertisement ad = AD_LIST.get(mIndexAdListHigh);

            PLAYER_MAP.entrySet().stream().filter(e -> e.getValue() == AdvertiseFrequency.HIGH).forEach(e -> e.getKey().sendMessage(PREFIX + ad.CONTENT));
        }, INTERVAL_TICK, INTERVAL_TICK);

        SCHEDULER.runTaskTimerAsynchronously(PLUGIN, () -> {
            if (AD_LIST.isEmpty() || PLAYER_MAP.isEmpty()) {
                return;
            }

            if (AD_LIST.size() <= ++mIndexAdListMiddle) {
                mIndexAdListMiddle = 0;
            }
            Advertisement ad = AD_LIST.get(mIndexAdListMiddle);

            PLAYER_MAP.entrySet().stream().filter(e -> e.getValue() == AdvertiseFrequency.MIDDLE).forEach(e -> e.getKey().sendMessage(PREFIX + ad.CONTENT));
        }, INTERVAL_TICK, INTERVAL_TICK * AdvertiseFrequency.MIDDLE.getIntervalGain());

        SCHEDULER.runTaskTimerAsynchronously(PLUGIN, () -> {
            if (AD_LIST.isEmpty() || PLAYER_MAP.isEmpty()) {
                return;
            }

            if (AD_LIST.size() <= ++mIndexAdListLow) {
                mIndexAdListLow = 0;
            }
            Advertisement ad = AD_LIST.get(mIndexAdListLow);

            PLAYER_MAP.entrySet().stream().filter(e -> e.getValue() == AdvertiseFrequency.LOW).forEach(e -> e.getKey().sendMessage(PREFIX + ad.CONTENT));
        }, INTERVAL_TICK, INTERVAL_TICK * AdvertiseFrequency.LOW.getIntervalGain());
    }

    public void addSetAdConfirmTask(Player player, String strPeriodDays, String content) {
        int periodDays;

        try {
            periodDays = Integer.parseInt(strPeriodDays);
        } catch (Exception e) {
            sendWarnMessage(player, strPeriodDays + "は整数ではありません.");
            return;
        }

        if (periodDays <= 0 || periodDays > 30) {
            sendWarnMessage(player, "日数は 1～30 の範囲で入力して下さい");
            return;
        }

        if (content.length() < 10) {
            sendWarnMessage(player, "広告の内容が短すぎます.");
            return;
        }

        if (content.length() > 250) {
            sendWarnMessage(player, "250文字を超える広告は登録できません.");
            return;
        }

        if (DAO.hasAdvertisement(player.getUniqueId())) {
            sendWarnMessage(player, "すでに広告が登録されています.");
            player.sendMessage(HINT_SET_AD_OVERLAP);
            return;
        }

        sendPreview(player, content);

        CONFIRM_TASK_MAP.remove(player);

        String setMsg = AD_SET_MESSAGE
                .replace("{days}", Integer.toString(periodDays))
                .replace("{price}", Integer.toString(PRICE_PER_DAY * periodDays));

        player.sendMessage(setMsg);
        player.sendMessage(CONFIRM_MESSAGE.replace("{action}", "登録"));

        CONFIRM_TASK_MAP.put(player, new ConfirmTask(ConfirmTask.Type.SET, content, periodDays));
        SCHEDULER.runTaskLaterAsynchronously(PLUGIN, () -> CONFIRM_TASK_MAP.remove(player), 20L * 30);
    }

    public void addUnsetAdConfirmTask(Player player) {
        Advertisement ad = DAO.getAdvertisement(player.getUniqueId());

        if (ad == null) {
            sendWarnMessage(player, "広告が登録されていません.");
            return;
        }

        sendPreview(player, ad.CONTENT);

        CONFIRM_TASK_MAP.remove(player);

        player.sendMessage(AD_UNSET_MESSAGE);
        player.sendMessage(CONFIRM_MESSAGE.replace("{action}", "消去"));

        CONFIRM_TASK_MAP.put(player, new ConfirmTask(ConfirmTask.Type.UNSET, null, 0));
        SCHEDULER.runTaskLaterAsynchronously(PLUGIN, () -> CONFIRM_TASK_MAP.remove(player), 20L * 30);
    }

    public void sendPreview(CommandSender sender, String content) {
        sender.sendMessage(PREFIX + ChatColor.translateAlternateColorCodes('&', content));
    }

    public void sendPlayerAdInfo(Player player) {
        Advertisement ad = DAO.getAdvertisement(player.getUniqueId());

        player.sendMessage(AD_INFO_INDEX);
        player.sendMessage(" 受信頻度: " + PLAYER_MAP.get(player).getName());
        player.sendMessage(" 登録広告: " + ((ad != null) ? ad.CONTENT : ChatColor.DARK_GRAY + "なし"));

        if (ad != null) {
            player.sendMessage(" 配信期限: " + ChatColor.RED + Util.formatDate(ad.EXPIRE_DATE) + ChatColor.GRAY + "まで");
        }
    }

    public void setAdvertiseFrequency(Player player, String strFreq) {
        AdvertiseFrequency freq;
        try {
            freq = AdvertiseFrequency.valueOf(strFreq.toUpperCase());
        } catch (Exception e) {
            player.sendMessage(PREFIX + ChatColor.YELLOW + strFreq + "は存在しません.");
            return;
        }

        if (PLAYER_CONFIG.saveAdvertiseFrequency(player.getUniqueId(), freq)) {
            PLAYER_MAP.put(player, freq);
            player.sendMessage(PREFIX + ChatColor.GREEN + "受信頻度を" + freq.getName() + ChatColor.GREEN + "に設定しました.");
        } else {
            player.sendMessage(PREFIX + ChatColor.YELLOW + "設定に失敗しました.");
        }
    }

    public void sendAdList(CommandSender sender) {
        sender.sendMessage(AD_LIST_INDEX);

        AD_LIST.forEach(ad -> sender.sendMessage(AD_LIST.indexOf(ad)+1 + ". " + ad.CONTENT + ChatColor.RESET + ChatColor.GRAY + " (" + ad.OWNER.getName() + ')'));
    }

    public void confirmTask(Player player) {
        ConfirmTask task = CONFIRM_TASK_MAP.get(player);

        if (task == null) {
            sendWarnMessage(player, "認証が必要な処理はありません");
            return;
        }

        CONFIRM_TASK_MAP.remove(player);

        switch (task.TYPE) {
            case SET:
                setPlayerAd(player, task.CONTENT, task.PERIOD_DAYS);
                break;
            case UNSET:
                unsetPlayerAd(player);
                break;
            default:
                break;
        }
    }

    public void sync() {
        syncLatestAds();
    }

    @Override
    public void onPlayerJoin(Player player) {
        PLAYER_MAP.put(player, PLAYER_CONFIG.getAdvertiseFrequency(player.getUniqueId()));
    }

    @Override
    public void onPlayerQuit(Player player) {
        PLAYER_MAP.remove(player);
    }

    private void setPlayerAd(Player player, String content, int periodDays) {
        if (content == null || periodDays < 1) {
            sendErrorMessage(player, "不正な操作です.");
            return;
        }

        double price = periodDays * PRICE_PER_DAY;
        double balance = ECONOMY.getBalance(player);
        if (balance < price) {
            sendWarnMessage(player, "所持金が \u00A5" + (price - balance) + " 不足しています.");
            return;
        }

        if (price < 0D || !ECONOMY.withdraw(player, price)) {
            sendErrorMessage(player, "購入処理に失敗しました.");
            return;
        }

        if (!DAO.addAdvertisement(player.getUniqueId(), content, ZonedDateTime.now().plusDays(periodDays))) {
            sendErrorMessage(player, "データベースの更新に失敗しました.");

            if (!ECONOMY.deposit(player, price)) {
                sendErrorMessage(player, "復元処理に失敗しました.");
            }
            return;
        }

        player.sendMessage(PREFIX + ChatColor.GREEN + "登録しました.");

        syncLatestAds();
    }

    private void unsetPlayerAd(Player player) {
        if (!DAO.hasAdvertisement(player.getUniqueId())) {
            sendWarnMessage(player, "広告が登録されていません.");
            return;
        }

        if (!DAO.deleteAdvertisement(player.getUniqueId())) {
            sendErrorMessage(player, "データベースの更新に失敗しました.");
            return;
        }

        player.sendMessage(PREFIX + ChatColor.GREEN + "消去しました.");

        syncLatestAds();
    }

    private void sendWarnMessage(CommandSender sender, String message) {
        sender.sendMessage(PREFIX + ChatColor.YELLOW + message);
    }

    private void sendErrorMessage(CommandSender sender, String message) {
        sender.sendMessage(PREFIX + ChatColor.RED + message);
    }

    private final String AD_LIST_INDEX = ChatColor.translateAlternateColorCodes('&', "&7***************&d 広告一覧&7 ***************");
    private final String AD_INFO_INDEX = ChatColor.translateAlternateColorCodes('&', "&7***************&d 広告情報&7 ***************");

    private final static String AD_SET_MESSAGE = ChatColor.translateAlternateColorCodes('&', PREFIX + "この広告を &a{days}日間 &e\u00A5{price} &rで登録しますか？");
    private final static String AD_UNSET_MESSAGE = ChatColor.translateAlternateColorCodes('&', PREFIX + "この広告を消去しますか？");
    private final static String CONFIRM_MESSAGE = ChatColor.translateAlternateColorCodes('&', PREFIX + "&7{action}を確定するには &c/ad confirm&7 と入力して下さい.");
    private final static String HINT_SET_AD_OVERLAP = ChatColor.translateAlternateColorCodes('&', PREFIX + "/ad unset で以前の広告を消去して下さい.");


    private static class ConfirmTask {
        enum Type {SET, UNSET}

        private final Type TYPE;
        private final String CONTENT;
        private final int PERIOD_DAYS;

        ConfirmTask(Type type, String content, int periodDays) {
            TYPE = type;
            CONTENT = content;
            PERIOD_DAYS = periodDays;
        }
    }
}

File: src/main/kotlin/jp/kentan/minecraft/nekocore/manager/AdvertisementManager.kt
package jp.kentan.minecraft.nekocore.manager

import jp.kentan.minecraft.nekocore.NekoCorePlugin
import jp.kentan.minecraft.nekocore.data.AdvertisementRepository
import jp.kentan.minecraft.nekocore.data.PlayerRepository
import jp.kentan.minecraft.nekocore.data.model.AdvertiseFrequency
import jp.kentan.minecraft.nekocore.data.model.Advertisement
import jp.kentan.minecraft.nekocore.util.dayToMills
import jp.kentan.minecraft.nekocore.util.formatYearMonthDayHm
import jp.kentan.minecraft.nekocore.util.sendDatabaseUpdateError
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.ArrayList

class AdvertisementManager(
    private val plugin: NekoCorePlugin
) {

    companion object {
        private const val PREFIX = "§7[§d広告§7]§r "
        private const val BASE_INTERVAL_TICK = 20L * 60 * 5 //5m
        private const val PRICE_PER_DAY = 100
    }

    private enum class Type { SET, UNSET }

    private val playerRepo = PlayerRepository(plugin)
    private val adRepo = AdvertisementRepository(plugin)

    private val advertisementList = Collections.synchronizedList(ArrayList<Advertisement>())
    private val playerMap: MutableMap<Player, AdvertiseFrequency> = ConcurrentHashMap()

    private val confirmTaskMap: MutableMap<Player, ConfirmTask> = ConcurrentHashMap()

    init {
        GlobalScope.launch {
            syncLatestAds()
            playerMap.putAll(Bukkit.getOnlinePlayers().map { it to playerRepo.getAdvertiseFrequency(it.uniqueId) })
        }

        plugin.bukkitEventListener.subscribePlayerJoin(::onPlayerJoined)
        plugin.bukkitEventListener.subscribePlayerQuit(::onPlayerQuited)

        startBroadcastAsyncTask()
    }

    private fun syncLatestAds() {
        GlobalScope.launch {
            advertisementList.clear()
            advertisementList.addAll(adRepo.getAdvertisementList())
        }
    }

    fun addSetAdConfirmTask(player: Player, strPeriodDays: String, contents: List<String>) {
        val periodDays = strPeriodDays.toIntOrNull() ?: let {
            player.sendMessage("$PREFIX${strPeriodDays}§eは整数ではありません.")
            return
        }

        if (periodDays !in 1..15) {
            player.sendMessage("${PREFIX}§e日数は 1～15 の範囲で入力してください.")
            return
        }

        val content: String = ChatColor.translateAlternateColorCodes('&', contents.joinToString(separator = " "))

        if (content.length < 10) {
            player.sendMessage("${PREFIX}§e広告の内容が短すぎます.")
            return
        }
        if (content.length > 140) {
            player.sendMessage("${PREFIX}§e140文字を超える広告は登録できません.")
            return
        }

        GlobalScope.launch {
            if (adRepo.hasAdvertisement(player.uniqueId)) {
                player.sendMessage("${PREFIX}§eすでに広告が登録されています.")
                player.sendMessage("$PREFIX/ad unset で以前の広告を消去して下さい.")
                return@launch
            }

            // preview
            player.sendMessage(PREFIX + content)

            confirmTaskMap.remove(player)

            player.sendMessage("${PREFIX}この広告を §a${periodDays}日間 §e\u00A5${PRICE_PER_DAY * periodDays} §rで登録しますか？")
            player.sendMessage("${PREFIX}§7登録を確定するには §c/ad confirm§7 と入力して下さい.")

            confirmTaskMap[player] = ConfirmTask(Type.SET, content, periodDays)
            Bukkit.getScheduler().runTaskLaterAsynchronously(plugin, Runnable { confirmTaskMap.remove(player) }, 20L * 30)
        }
    }

    fun addUnsetAdConfirmTask(player: Player) {
        GlobalScope.launch {
            val ad = adRepo.getAdvertisement(player.uniqueId) ?: let {
                player.sendMessage("${PREFIX}§e広告が登録されていません.")
                return@launch
            }

            // preview
            player.sendMessage(PREFIX + ad.content)

            confirmTaskMap.remove(player)

            player.sendMessage("${PREFIX}この広告を消去しますか？")
            player.sendMessage("${PREFIX}§7消去を確定するには §c/ad confirm§7 と入力して下さい.")

            confirmTaskMap[player] = ConfirmTask(Type.UNSET, "", 0)
            Bukkit.getScheduler().runTaskLaterAsynchronously(plugin, Runnable { confirmTaskMap.remove(player) }, 20L * 30)
        }
    }

    fun sendInfo(player: Player) {
        GlobalScope.launch {
            val ad = adRepo.getAdvertisement(player.uniqueId)

            player.sendMessage("§7***************§d 広告情報§7 ***************")
            player.sendMessage(" 受信頻度: ${playerMap[player]!!.displayName}")
            player.sendMessage(" 登録広告: ${ad?.content ?: "§8なし"}")
            if (ad != null) {
                player.sendMessage(" 配信期限: §c${ad.expiredDate.formatYearMonthDayHm()}§7まで")
            }
        }
    }

    fun sendList(sender: CommandSender) {
        sender.sendMessage("§7***************§d 広告一覧§7 ***************")
        advertisementList.forEachIndexed { index, ad ->
            sender.sendMessage("${index + 1}. ${ad.content} §7(${ad.owner.name})")
        }
    }

    fun preview(sender: CommandSender, contents: List<String>) {
        val content: String = ChatColor.translateAlternateColorCodes('&', contents.joinToString(separator = " "))
        sender.sendMessage(PREFIX + content)
    }

    fun setAdvertiseFrequency(player: Player, strFreq: String) {
        val freq = try {
            AdvertiseFrequency.valueOf(strFreq.toUpperCase())
        } catch (e: Exception) {
            player.sendMessage("${PREFIX}§e${strFreq}は存在しません.")
            return
        }

        GlobalScope.launch {
            if (playerRepo.updateAdvertiseFrequency(player.uniqueId, freq)) {
                playerMap[player] = freq
                player.sendMessage("${PREFIX}§a受信頻度を${freq.displayName}§aに設定しました.")
            } else {
                player.sendMessage("${PREFIX}§e設定に失敗しました.")
            }
        }
    }

    fun confirmTask(player: Player) {
        confirmTaskMap.remove(player)?.run(player) ?: let {
            player.sendMessage("${PREFIX}§e認証が必要な処理はありません.")
            return
        }
    }

    fun sync(sender: CommandSender) {
        syncLatestAds()
        sender.sendMessage("${PREFIX}§a最新のデータベースと同期しました.")
    }

    private fun startBroadcastAsyncTask() {
        var indexHigh = 0
        var indexMiddle = 0
        var indexLow = 0

        Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, Runnable {
            val now = System.currentTimeMillis()
            advertisementList.removeIf { it.expiredDate.time <= now }
            if (advertisementList.isEmpty()) {
                return@Runnable
            }

            if (advertisementList.size <= ++indexHigh) {
                indexHigh = 0
            }
            broadcast(AdvertiseFrequency.HIGH, indexHigh)
        }, BASE_INTERVAL_TICK, BASE_INTERVAL_TICK * AdvertiseFrequency.HIGH.intervalGain)

        Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, Runnable {
            if (advertisementList.isEmpty()) {
                return@Runnable
            }

            if (advertisementList.size <= ++indexMiddle) {
                indexMiddle = 0
            }
            broadcast(AdvertiseFrequency.MIDDLE, indexMiddle)
        }, BASE_INTERVAL_TICK, BASE_INTERVAL_TICK * AdvertiseFrequency.MIDDLE.intervalGain)

        Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, Runnable {
            if (advertisementList.isEmpty()) {
                return@Runnable
            }

            if (advertisementList.size <= ++indexLow) {
                indexLow = 0
            }
            broadcast(AdvertiseFrequency.LOW, indexLow)
        }, BASE_INTERVAL_TICK, BASE_INTERVAL_TICK * AdvertiseFrequency.LOW.intervalGain)
    }

    private fun broadcast(freq: AdvertiseFrequency, index: Int) {
        val ad = advertisementList[index]
        val message = "$PREFIX${ad.content}"

        playerMap.filterValues { it == freq }
            .forEach { (player, _) -> player.sendMessage(message) }
    }

    private fun onPlayerJoined(player: Player) {
        GlobalScope.launch { playerMap[player] = playerRepo.getAdvertiseFrequency(player.uniqueId) }
    }

    private fun onPlayerQuited(player: Player) {
        playerMap.remove(player)
    }

    private inner class ConfirmTask(
        val type: Type,
        val content: String,
        val periodDays: Int
    ) {

        fun run(player: Player) {
            GlobalScope.launch {
                when (type) {
                    Type.SET -> runSetTask(player)
                    Type.UNSET -> runUnsetTask(player)
                }
            }
        }

        private fun runSetTask(player: Player) {
            val price = periodDays * PRICE_PER_DAY
            val balance = plugin.economy.getBalance(player)

            if (balance < price) {
                player.sendMessage("${PREFIX}§e所持金が \u00A5${price - balance} 不足しています.")
                return
            }

            if (!plugin.economy.withdrawPlayer(player, price.toDouble()).transactionSuccess()) {
                player.sendMessage("${PREFIX}§e購入処理に失敗しました.")
                return
            }

            val expiredDate = Date(System.currentTimeMillis() + (periodDays.dayToMills()))

            if (!adRepo.addAdvertisement(player.uniqueId, content, expiredDate)) {
                player.sendDatabaseUpdateError()

                plugin.economy.depositPlayer(player, price.toDouble())
                return
            }

            player.sendMessage("${PREFIX}§a登録しました.")

            syncLatestAds()
        }

        private fun runUnsetTask(player: Player) {
            if (!adRepo.deleteAdvertisement(player.uniqueId)) {
                player.sendDatabaseUpdateError()
                return
            }

            player.sendMessage("${PREFIX}§a消去しました.")

            syncLatestAds()
        }
    }
}